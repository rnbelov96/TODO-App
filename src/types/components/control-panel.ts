import FilterStatus from '@/contstants';
import { TODOType } from '../general-types';

export type ControlPanelPropsType = {
  unfinishedTODOCount: number;
  filterBy: FilterStatus;
  todos: TODOType[];
  onSwitch: (value: FilterStatus) => void;
  onDeleteCompleted: () => void;
};
