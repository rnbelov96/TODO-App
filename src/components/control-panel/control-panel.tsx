import FilterStatus from '@/contstants';
import { AppActionCreators } from '@/redux/app';
import {
  getFilterStatus,
  getTODOSToShow,
  getUnfinishedTODOCount,
} from '@/redux/app/selectors';
import { ControlPanelPropsType } from '@/types/components/control-panel';
import { FullStateType } from '@/types/general-types';
import { AppActionType } from '@/types/redux/app-reducer';
import React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

export const PureControlPanel: React.FunctionComponent<ControlPanelPropsType> = ({
  unfinishedTODOCount,
  filterBy,
  todos,
  onSwitch,
  onDeleteCompleted,
}) => (
  <div className="controlPanel">
    <div
      className="controlPanel__itemsLeft"
      style={unfinishedTODOCount === 0 ? { opacity: 0 } : { opacity: 1 }}
    >
      {unfinishedTODOCount}
      {' '}
      items left
    </div>
    <div className="controlPanel__groupButtons">
      <button
        type="button"
        className={
          filterBy === FilterStatus.ALL
            ? 'controlPanel__groupButton controlPanel__groupButton_active'
            : 'controlPanel__groupButton'
        }
        onClick={() => onSwitch(FilterStatus.ALL)}
      >
        All
      </button>
      <button
        type="button"
        className={
          filterBy === FilterStatus.ACTIVE
            ? 'controlPanel__groupButton controlPanel__groupButton_active'
            : 'controlPanel__groupButton'
        }
        onClick={() => onSwitch(FilterStatus.ACTIVE)}
      >
        Active
      </button>
      <button
        type="button"
        className={
          filterBy === FilterStatus.COMPLETED
            ? 'controlPanel__groupButton controlPanel__groupButton_active'
            : 'controlPanel__groupButton'
        }
        onClick={() => onSwitch(FilterStatus.COMPLETED)}
      >
        Completed
      </button>
    </div>
    <button
      type="button"
      className="controlPanel__clearButton"
      style={
        todos.filter(el => el.done === true).length
          ? { opacity: 1 }
          : { opacity: 0 }
      }
      onClick={onDeleteCompleted}
    >
      Clear completed
    </button>
  </div>
);

function mapStateToProps(state: FullStateType) {
  return {
    filterBy: getFilterStatus(state),
    unfinishedTODOCount: getUnfinishedTODOCount(state),
    todos: getTODOSToShow(state),
  };
}

function mapDispatchToProps(dispatch: Dispatch<AppActionType>) {
  return {
    onSwitch: (value: FilterStatus) => dispatch(AppActionCreators.switchFilter(value)),
    onDeleteCompleted: () => dispatch(AppActionCreators.deleteCompletedTODO()),
  };
}

export const ControlPanel = connect(mapStateToProps, mapDispatchToProps)(PureControlPanel);
