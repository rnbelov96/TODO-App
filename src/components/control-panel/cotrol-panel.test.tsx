import FilterStatus, { mockFullInitiateState } from '@/contstants';
import React from 'react';
import renderer from 'react-test-renderer';
import { PureControlPanel } from './control-panel';

describe('ControlPanel', () => {
  test('should render control panel', () => {
    const tree = renderer
      .create(
        <PureControlPanel
          filterBy={FilterStatus.ALL}
          onDeleteCompleted={() => {}}
          onSwitch={() => {}}
          todos={mockFullInitiateState.app.todos}
          unfinishedTODOCount={1}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
