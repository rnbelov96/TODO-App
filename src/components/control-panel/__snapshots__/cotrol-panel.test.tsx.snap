// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanel should render control panel 1`] = `
<div
  className="controlPanel"
>
  <div
    className="controlPanel__itemsLeft"
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    1
     
    items left
  </div>
  <div
    className="controlPanel__groupButtons"
  >
    <button
      className="controlPanel__groupButton controlPanel__groupButton_active"
      onClick={[Function]}
      type="button"
    >
      All
    </button>
    <button
      className="controlPanel__groupButton"
      onClick={[Function]}
      type="button"
    >
      Active
    </button>
    <button
      className="controlPanel__groupButton"
      onClick={[Function]}
      type="button"
    >
      Completed
    </button>
  </div>
  <button
    className="controlPanel__clearButton"
    onClick={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    type="button"
  >
    Clear completed
  </button>
</div>
`;
